### PR TITLE
GH-33804: [Python] Add support for manylinux_2_28 wheel

### LINKED ISSUE #33804: [Packaging][Python] Add support for manylinux_2_28 wheels
### Describe the enhancement requested

This is low priority as I'm not on the PMC or a Committer. However, I thought I'd create it as I wanted to create a pyarrow wheel with the new C++ ABI: `_GLIBCXX_USE_CXX11_ABI=1`. In the process of doing so, I created a manylinux_2_28 wheel by adapting the existing manylinux2014 Dockerfile which may prove useful.

Related:

- https://pypackaging-native.github.io/key-issues/native-dependencies/cpp_deps/
- https://github.com/apache/arrow/issues/32415

I'll submit the manylinux_2_28 Dockerfile in a PR supporting this enhancement.

### Component(s)

Python

### PR DESCRIPTION
**Pull request by sjperkins**
Closes #33804 

### Rationale for this change

At some point, it would be useful to support the new C++ ABI `_GLIBCXX_USE_CXX11_ABI=1` in pyarrow wheels, especially when moving to C++17:

- https://github.com/apache/arrow/issues/32415

I wanted to create a pyarrow wheel that supported the above ABI and adapted the existing CENTOS 7 manylinux2014 Dockerfile/wheel to produce a AlmaLinux 8 manylinux_2_28 Dockerfile/wheel.

Publishing wheels with a new ABI needs [careful consideration](https://pypackaging-native.github.io/key-issues/native-dependencies/cpp_deps/) so I think this is low priority, but I thought I'd provide this manylinux_2_28 implementation in case it was useful for current/future adoption. 

### What changes are included in this PR?

A manylinux_2_28 Dockerfile, adopted from the existing manylinux2014 Dockerfile

### Are these changes tested?

Manually tested at present

### Are there any user-facing changes?

Yes, there's a major ABI change, as pyarrow will be compiled with `_GLIBCXX_USE_CXX11_ABI=1`
* Closes: #33804

Supercedes:
* https://github.com/apache/arrow/pull/33805
* Closes: #33804

### CODE REVIEW

**Comment by liujiajun** [comment]
Hi I am new to Arrow and I would like to ask a dumb question here:

From my understanding, the current situation (before this PR) is 
- pyarrow is compiled with the old ABI
- arrow is compile with the newer ABI. This can be verified by `nm -C libarrow.so.1100 | grep cxx11`

Then, how come pyarrow is able to use arrow without a runtime error? 

Thanks!

**Comment by kou** [comment]
@github-actions crossbow submit wheel-manylinux*

**Comment by sjperkins** [comment]
> Then, how come pyarrow is able to use arrow without a runtime error?

The pyarrow binary wheel packages the arrow shared object libraries built with the ABI, and I believe auditwheel correctly configures the linking to these shared objects.

**Comment by sjperkins** [comment]
linux-headers issue still occurring on arm-64

As mentioned in #34818 looks like this could be caused by:

- https://github.com/microsoft/vcpkg/issues/29674

which could be resolved by

- https://github.com/microsoft/vcpkg/pull/29966
- https://github.com/microsoft/vcpkg/issues/29674#issuecomment-1459940460

Maybe a newer release of vcpkg is required? Pushing this forward takes an excessive amount of time for me locally due to arm64 emulation.

**Comment by kou** [comment]
Could you try more recent vcpkg and run `@github-actions crossbow submit wheel-manylinux-2-28-cp310-arm64`?
You don't need to try it on local for this.

**Comment by kou** [comment]
Could you rebase on main to use #34835?

**Comment by sjperkins** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp310-arm64

**Comment by sjperkins** [comment]
> Could you try more recent vcpkg and run `@github-actions crossbow submit wheel-manylinux-2-28-cp310-arm64`? You don't need to try it on local for this.

Thanks, the merge and cross submit seems to work. Will try a more recent vcpkg during the week.

**Comment by sjperkins** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp310-arm64

**Comment by kou** [comment]
Could you rebase on main for the pandas related failures?
Recent pandas related fix may solve the error.

**Comment by sjperkins** [comment]
> Could you rebase on main for the pandas related failures? Recent pandas related fixes may solve the error.

Done. Looks like `linux-headers` issues in arm64 didn't occur in https://github.com/ursacomputing/crossbow/actions/runs/4621955303/jobs/8174022199.

**Comment by sjperkins** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp310-arm64

**Comment by sjperkins** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp310-arm64

**Comment by sjperkins** [comment]
Green amd64 build 😄 . Thanks for updating to LLVM 16 @kou too -- that fixed the MINGW issues.

**Comment by sjperkins** [comment]
@github-actions crossbow submit wheel-manylinux*

**Comment by sjperkins** [comment]
@github-actions crossbow submit wheel-manylinux-2-28-cp38*

**Comment by kou** [comment]
I re-run the job.

File: .env
Comment (sjperkins) [inline .env:102]: The above isn't a released version (but does contain the openssl linux-headers fixes). Would it make sense to wait for a new release? vcpkg seems to release every 2 months.

**Review by sjperkins — COMMENTED** [review]
Looks like the re-run avoided https://github.com/apache/arrow/issues/15054 so in principle it looks like the wheels are OK. @kou do you think it's worth re-running the entire manylinux wheel crossbow action again?

What else would need to be done?

File: .env
Comment (kou) [reply .env:102]: Could you open a new issue to update a release version and update the TODO comment to refer the new issue?

File: .env
Comment (kou) [inline .env:100]: Could you remove this?
This is needless, right?

**Comment by kou** [comment]
@github-actions crossbow submit wheel-windows-*

File: .env
Comment (sjperkins) [reply .env:102]: Yes, have done so in https://github.com/apache/arrow/issues/35099

File: .env
Comment (sjperkins) [reply .env]: I have removed it

**Comment by sjperkins** [comment]
> Could you rebase on main? I want to test "java-jars" jobs because it also uses vcpkg.

done

**Comment by kou** [comment]
Thanks! But sorry... Please one more rebase for #35092...

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by sjperkins** [comment]
> Revision: 22bb08f68880baf1683b9f5d1d7bdf98c221084b
> 
> Submitted crossbow builds: [ursacomputing/crossbow @ actions-dc1ed1ccb8](https://github.com/ursacomputing/crossbow/branches/all?query=actions-dc1ed1ccb8)
> 
> |Task|Status|
> |----|------|
> |java-jars|[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/4695407611/jobs/8324549481)|

Looks promising. Should I rebase on main again?

**Comment by kou** [comment]
Not needed but could you merge this after we release 12.0.0?
If we merge this immediately, we can't use the current vcpkg cache to build 12.0.0 artifacts.

**Review by kou — APPROVED** [review]

**Comment by kou** [comment]
The 12.0.0 vote passed.
So I merge this.

File: .env
Comment (kou) [inline .env:98]: Oh, it seems that this is outdated.
https://github.com/ursacomputing/crossbow/pull/7 automated this.

Comment (sjperkins) [reply .env:98]: I removed this comment in https://github.com/apache/arrow/pull/35430. I suggest continuing the discussion there?

**Pull request merged**
